Sources/SecureXPC/Routes.swift

```diff
@@ -7,8 +7,57 @@
 
 import Foundation
 
-/// Consistent framework internal implementation of routes that can be sent over XPC (because its Codable) and used as a dictionary key (because its Hashable).
-struct XPCRoute: Codable, Hashable {
+/// The entry point for creating routes needed for client to server communication.
+///
+/// Client to server communication is performed using routes. A route is a sequence of `String`s and if applicable also the message and reply types.
+///
+/// In practice a route is similar to a function signature or a server path with type safety, although it is not precisely either. Message and reply types must be
+/// [`Codable`](https://developer.apple.com/documentation/swift/codable). Many structs, enums, and classes in the Swift
+/// standard library are already `Codable` and compiler generated conformance is available for simple structs and enums.
+///
+/// The simplest form of a route is one that contains neither a message nor a reply:
+/// ```swift
+/// let resetRoute = XPCRoute.named("reset")
+/// ```
+///
+/// In many cases a reply will be desired:
+/// ```swift
+/// let thermalLastSetRoute = XPCRoute.named("thermal", "lastSetAt")
+///                                   .withReplyType(Date.self)
+/// ```
+///
+/// Routes can have a message sent to them that don't have a reply:
+/// ```swift
+/// let setLimitRoute = XPCRoute.named("thermal", "setLimit")
+///                             .withMessageType(Int.self)
+/// ```
+///
+/// As well as those that do expect a reply:
+/// ```swift
+/// let setLimitRoute = XPCRoute.named("thermal", "setLimit")
+///                             .withMessageType(Int.self)
+///                             .withReplyType(Int.self)
+/// ```
+///
+/// Routes are distinct based on their names, meaning the last two routes above are considered equivalent. A server will
+/// only allow one handler to be registered for each distinct route.
+///
+/// #### Storing Routes
+/// To ensure consistency, ideally routes are only defined once in code that is shared by both the client and server.
+///
+/// #### Registering Routes
+/// Once a route has been created, it needs to be registered with an ``XPCServer`` in order for it to be called by an ``XPCClient``.
+///
+/// #### Updating Routes
+/// If you are using XPC Mach services, take care when updating existing routes because over time you may end up with an
+/// older version of your server installed on a computer with a newer client. You may want to version your routes by
+/// prefixing them with a number:
+/// ```swift
+/// XPCRoute.named("1", "reset")
```

Comment:
Brain-storming: once we have more real world experience maintaining these, perhaps we should build versioning/fallbacks/etc. as a first class feature of the framework

Reply:
The thought I've had in this space is to create built-in support for a client to query if a server has a certain route registered.

Reply:
I don't think that would be sufficient, because the name/msg-type/reply-type could all be the same, but they might still be incompatible due to changes in the msg/reply types.

Now that I think abotu it, that's more of a "Codable doesn't have versioning facilities" problem, not necessarily specific to SecureXPC.

Reply:
Yeah checking for registered routes wouldn't be something automatic, but combined with proper versioning could work well.  Quite possibly there are better ways to achieve this. 

In terms of `Codable` the issue is that actually the types don't need to match across client and server for it to be valid. As long as encoding/decoding is possible, it's fine. And in some ways this is a good thing for example if the type is renamed, but not otherwise changed. I don't know there's a way to validate an encoding/decoding is possible without actually trying it. Haven't given this much thought though.

Precise versioning control is why things like gRPC exist. (I actually briefly contemplated using gRPC instead of `Codable`.)